File: driver_app/lib/models/ride.dart
/// Shared ride model — mirrors the rider app's Ride model so both apps
/// speak the same backend schema.

class LocationPoint {
  final double lat;
  final double lng;

  const LocationPoint({required this.lat, required this.lng});

  factory LocationPoint.fromJson(Map<String, dynamic> json) => LocationPoint(
        lat: (json['lat'] as num).toDouble(),
        lng: (json['lng'] as num).toDouble(),
      );

  Map<String, dynamic> toJson() => {'lat': lat, 'lng': lng};
}

enum RideStatus {
  requested,
  matched,
  driverEnRoute,
  arrived,
  inProgress,
  completed,
  cancelled;

  static RideStatus fromString(String s) {
    switch (s) {
      case 'requested':
        return requested;
      case 'matched':
        return matched;
      case 'driver_en_route':
        return driverEnRoute;
      case 'arrived':
        return arrived;
      case 'in_progress':
        return inProgress;
      case 'completed':
        return completed;
      case 'cancelled':
        return cancelled;
      default:
        throw ArgumentError('Unknown ride status: $s');
    }
  }

  String toApiString() {
    switch (this) {
      case requested:
        return 'requested';
      case matched:
        return 'matched';
      case driverEnRoute:
        return 'driver_en_route';
      case arrived:
        return 'arrived';
      case inProgress:
        return 'in_progress';
      case completed:
        return 'completed';
      case cancelled:
        return 'cancelled';
    }
  }
}

class Ride {
  final int id;
  final RideStatus status;
  final String pickupAddress;
  final String dropoffAddress;
  final double estimatedFare;
  final double? actualFare;
  final String? riderName;
  final String? riderPhone;
  final LocationPoint? pickupLocation;
  final LocationPoint? dropoffLocation;
  final double? distanceKm;
  final double? durationMin;
  final DateTime requestedAt;
  final DateTime? matchedAt;
  final DateTime? startedAt;
  final DateTime? completedAt;

  const Ride({
    required this.id,
    required this.status,
    required this.pickupAddress,
    required this.dropoffAddress,
    required this.estimatedFare,
    this.actualFare,
    this.riderName,
    this.riderPhone,
    this.pickupLocation,
    this.dropoffLocation,
    this.distanceKm,
    this.durationMin,
    required this.requestedAt,
    this.matchedAt,
    this.startedAt,
    this.completedAt,
  });

  factory Ride.fromJson(Map<String, dynamic> json) => Ride(
        id: json['id'] as int,
        status: RideStatus.fromString(json['status'] as String),
        pickupAddress: json['pickup_address'] as String,
        dropoffAddress: json['dropoff_address'] as String,
        estimatedFare: (json['estimated_fare'] as num).toDouble(),
        actualFare: (json['actual_fare'] as num?)?.toDouble(),
        riderName: json['rider_name'] as String?,
        riderPhone: json['rider_phone'] as String?,
        pickupLocation: json['pickup'] != null
            ? LocationPoint.fromJson(json['pickup'] as Map<String, dynamic>)
            : null,
        dropoffLocation: json['dropoff'] != null
            ? LocationPoint.fromJson(json['dropoff'] as Map<String, dynamic>)
            : null,
        distanceKm: (json['distance_km'] as num?)?.toDouble(),
        durationMin: (json['duration_min'] as num?)?.toDouble(),
        requestedAt: DateTime.parse(json['requested_at'] as String),
        matchedAt: json['matched_at'] != null
            ? DateTime.parse(json['matched_at'] as String)
            : null,
        startedAt: json['started_at'] != null
            ? DateTime.parse(json['started_at'] as String)
            : null,
        completedAt: json['completed_at'] != null
            ? DateTime.parse(json['completed_at'] as String)
            : null,
      );

  Ride copyWith({
    RideStatus? status,
    double? actualFare,
    DateTime? matchedAt,
    DateTime? startedAt,
    DateTime? completedAt,
  }) =>
      Ride(
        id: id,
        status: status ?? this.status,
        pickupAddress: pickupAddress,
        dropoffAddress: dropoffAddress,
        estimatedFare: estimatedFare,
        actualFare: actualFare ?? this.actualFare,
        riderName: riderName,
        riderPhone: riderPhone,
        pickupLocation: pickupLocation,
        dropoffLocation: dropoffLocation,
        distanceKm: distanceKm,
        durationMin: durationMin,
        requestedAt: requestedAt,
        matchedAt: matchedAt ?? this.matchedAt,
        startedAt: startedAt ?? this.startedAt,
        completedAt: completedAt ?? this.completedAt,
      );
}

File: backend/app/models/ride.py
import enum
from datetime import datetime

from geoalchemy2 import Geometry
from sqlalchemy import DateTime, Enum, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base
from app.models.vehicle import VehicleServiceCategory


class RideStatus(str, enum.Enum):
    SCHEDULED = "scheduled"
    REQUESTED = "requested"
    MATCHED = "matched"
    DRIVER_EN_ROUTE = "driver_en_route"
    ARRIVED = "arrived"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class Ride(Base):
    __tablename__ = "rides"

    id: Mapped[int] = mapped_column(primary_key=True)
    rider_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True)
    driver_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True, index=True)
    status: Mapped[RideStatus] = mapped_column(Enum(RideStatus), default=RideStatus.REQUESTED)

    pickup_location: Mapped[bytes] = mapped_column(Geometry(geometry_type="POINT", srid=4326))
    dropoff_location: Mapped[bytes] = mapped_column(Geometry(geometry_type="POINT", srid=4326))
    pickup_address: Mapped[str] = mapped_column(String(500))
    dropoff_address: Mapped[str] = mapped_column(String(500))

    estimated_fare: Mapped[float]
    actual_fare: Mapped[float | None] = mapped_column(nullable=True)
    distance_km: Mapped[float | None] = mapped_column(nullable=True)
    duration_min: Mapped[float | None] = mapped_column(nullable=True)
    tip_amount: Mapped[float] = mapped_column(default=0.0)
    promo_code_id: Mapped[int | None] = mapped_column(nullable=True, index=True)
    promo_discount: Mapped[float] = mapped_column(default=0.0)

    scheduled_for: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    matched_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    cancelled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    is_pool: Mapped[bool] = mapped_column(default=False)
    pool_id: Mapped[int | None] = mapped_column(ForeignKey("ride_pools.id"), nullable=True, index=True)
    accessibility_required: Mapped[bool] = mapped_column(default=False)
    vehicle_type_preference: Mapped[VehicleServiceCategory | None] = mapped_column(
        Enum(VehicleServiceCategory), nullable=True
    )

    dispatch_retry_count: Mapped[int] = mapped_column(default=0)
    last_retry_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Reminder tracking — prevents duplicate reminders for scheduled rides
    reminder_1hr_sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    reminder_15min_sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # If generated from a recurring ride template
    recurring_ride_id: Mapped[int | None] = mapped_column(
        ForeignKey("recurring_rides.id"), nullable=True, index=True
    )

    rider_rating: Mapped[int | None] = mapped_column(nullable=True)
    driver_rating: Mapped[int | None] = mapped_column(nullable=True)
    cancellation_reason: Mapped[str | None] = mapped_column(Text, nullable=True)

    rider = relationship("User", foreign_keys=[rider_id], backref="rides_as_rider")
    driver = relationship("User", foreign_keys=[driver_id], backref="rides_as_driver")
